### PR TITLE
[Feature] Parse avro data directly

### PR DESCRIPTION
**Pull request by zaorangyang**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Previously, we used json to parse avro data #18265, but the performance is not good. This PR uses avro SDK to parse avro data directly.
The following is the test result by selecting the data of 10000000 lines before clickbench dataset. Note that this test does not include disk IO and data flopping.
+ JsonScanner: 109s
+ AvroScanner: 108s
+ AvroScanner-old: 826s

As you can see, this PR results in a nearly 8X improvement over the previous implementation.
In addition, we also tested the scenario of routine load importing avro data, which is a clickbench data set with a total of 99997497 lines of data.  The clusters tested are: single kafka broker + 5 partitions, single BE node, clickbench table 8 buckets, duplicate model. The test results are as follows:
- Avro: 36231 lines/s
- Json: 58753 lines/s

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [x] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/avro_scanner.cpp [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+    }
+    return Status::OK();
+}
+
+// This function extracts the corresponding field data from avro data according to
+// the corresponding path of each column, where:
+// input_value: input param
+// paths: input param
+// output_value: output param
+Status AvroScanner::_extract_field(const avro_value_t& input_value, std::vector<AvroPath> paths,
```
Comment (chaoyli) [inline be/src/exec/avro_scanner.cpp:617]: use `const std::vector<AvroPath>& paths` instead.

File: be/src/exec/avro_scanner.cpp [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+        } else if (st.is_not_found()) {
+            column->append_nulls(1);
+        } else {
+            return st;
+        }
+    }
+    return Status::OK();
+}
+
+Status AvroScanner::_parse_avro(Chunk* chunk, std::shared_ptr<SequentialFile> file) {
```
Comment (chaoyli) [inline be/src/exec/avro_scanner.cpp:241]: const std::shared_ptr<SequentialFile>&

File: be/src/exec/avro_scanner.cpp [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+
+StatusOr<ChunkPtr> AvroScanner::get_next() {
+    SCOPED_RAW_TIMER(&_counter->total_ns);
+    ChunkPtr src_chunk;
+    RETURN_IF_ERROR(_create_src_chunk(&src_chunk));
+    const int chunk_capacity = _state->chunk_size();
+    src_chunk->reserve(chunk_capacity);
+    src_chunk->set_num_rows(0);
+    Status st = _parse_avro(src_chunk.get(), _file);
+    if (!st.ok()) {
```
Comment (chaoyli) [inline be/src/exec/avro_scanner.cpp:395]:
```suggestion
    if (!st.ok() && !st.is_time_out() && !st.is_end_of_file()) {
```

File: be/src/exec/avro_scanner.h [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    Status _construct_column(const avro_value_t& input_value, Column* column, const TypeDescriptor& type_desc,
+                             const std::string& col_name);
+    Status _extract_field(const avro_value_t& input_value, std::vector<AvroPath> paths, avro_value_t& output_value);
+    Status _handle_union(avro_value_t input_value, avro_value_t& branch);
+    Status _get_array_element(avro_value_t* cur_value, size_t idx, avro_value_t* element);
+    std::string _preprocess_jsonpaths(std::string jsonpath);
+    Status _construct_row_without_jsonpath(const avro_value_t& avro_value, Chunk* chunk);
+
+    const TBrokerScanRange& _scan_range;
+    serdes_t* _serdes;
```
[the remark this answers is not in the document]
Comment (zaorangyang) [reply be/src/exec/avro_scanner.h:87]: > You can use `std::shared_ptr<serders_t*>` to replace the naked pointer.
serders_t is a incomplete type from external libserdes, we cannot get the full declaration. Please see: 
https://stackoverflow.com/questions/17649882/shared-ptr-with-incomplete-types-from-library

File: be/src/exec/avro_scanner.cpp [hunk 5]
```diff
@@ -0,0 +1,721 @@
...
+#include "gutil/casts.h"
+#include "gutil/strings/substitute.h"
+#include "runtime/exec_env.h"
+#include "runtime/runtime_state.h"
+#include "runtime/stream_load/load_stream_mgr.h"
+#include "runtime/stream_load/stream_load_pipe.h"
+#include "runtime/types.h"
+#include "util/defer_op.h"
+#include "util/runtime_profile.h"
+#ifdef __cplusplus
```
Comment (rickif) [inline be/src/exec/avro_scanner.cpp:49]: Could this macro be removed ?

**Review by chaoyli — APPROVED** [review]

**Review by meegoo — APPROVED** [review]

**Pull request merged**